lesson1_decorators/Homework/homework1.py

```diff
@@ -0,0 +1,71 @@
+from heapq import nsmallest
+from operator import itemgetter
+import functools
+import requests
+
+
+# TODO: 1. Реалізувати LFU алгоритм для кешування. За базу берем існуючий декоратор.Написати для фетчування URL-ів.
+#  Додати можливість указати максимум елементів в кеші.
+
+class Counter(dict):
+    # Відображає, де значення за замовчуванням дорівнюють нулю
+    def __missing__(self, key):
+        return 0
+
+
+def lfu_cache(maxsize=100):
+    def decorating_function(func):
+        cache = {}  # зіставлення аргументів з результатами
+        use_count = Counter()  # кількість звернень до кожного ключа
+        kwd_mark = object()  # окремі позиційні та ключові аргументи
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            key = args
+            if kwargs:
+                key += (kwd_mark,) + tuple(sorted(kwargs.items()))
+            use_count[key] += 1
+            # отримати запис кешу або створити, якщо не знайдено
+            try:
+                result = cache[key]
+                wrapper.hits += 1
+            except KeyError:
+                result = func(*args, **kwargs)
+                cache[key] = result
+                wrapper.misses += 1
+                # видаляємо найменш використовуваний запис в пам'яті
+                if len(cache) > maxsize:
+                    for key, _ in nsmallest(maxsize // 10,
```

Comment:
You dont need for, try to remove it without for loop.

Reply:
Done